src/codegen/compile.ml

```diff
@@ -7910,6 +7918,8 @@ module FuncDec = struct
         Lifecycle.trans env Lifecycle.Idle
       | Type.Shared Type.Query ->
         Lifecycle.trans env Lifecycle.PostQuery
+      | Type.Shared Type.Composite ->
+        Lifecycle.trans env Lifecycle.Idle
```

Comment:
Just to double check: The subsequent state of the lifecycle is `Idle` and not `PostQuery` for composite queries?

Reply:
Yes, this is an unfortunate hack. For queries, you can just generate the atomic body and enter PostQuery. But composite queries are not atomic, so you need to enter some state after every callback. Unfortunately, the code for generating callbacks always enters InUpdate and its hard to make the transition conditional.
I couldn't see any (easy) way to enter a different state without complicated the rest of the code to track whether we are in an ordinary callback or a composite callback. I could may set and test a global but that seemed gross too.

As far as I know, the Lifecycle stuff is just for sanity checking, so I think we are just losing some accuracy here.